From 5ac7347f79176f59643d313fdee71ea0bb7a60ed Mon Sep 17 00:00:00 2001
From: Bryan Quigley <bryan.quigley@canonical.com>
Date: Mon, 27 Nov 2017 00:08:14 -0500
Subject: [PATCH] Add generated systemd unit

Both Fedora and Debian have their own systemd units for boinc;
this is based on elements of both so we stop duplicating effort.

Also added clean to init.d script and made it only install if a
init.d directory already exists.

Closes: #2255
---
 .gitignore                             |  1 +
 client/scripts/Makefile.am             | 14 ++++++++++++--
 client/scripts/boinc-client.service.in | 21 +++++++++++++++++++++
 configure.ac                           |  1 +
 4 files changed, 35 insertions(+), 2 deletions(-)
 create mode 100644 client/scripts/boinc-client.service.in

diff --git a/.gitignore b/.gitignore
index a7745b5cf7..f32845e326 100644
--- a/.gitignore
+++ b/.gitignore
@@ -81,6 +81,7 @@ boinc_path_config.py
 pkginfo
 prototype
 client/scripts/boinc-client
+client/scripts/boinc-client.service
 config.h
 config.log
 config.status
diff --git a/client/scripts/Makefile.am b/client/scripts/Makefile.am
index 67d3820156..92153e6835 100644
--- a/client/scripts/Makefile.am
+++ b/client/scripts/Makefile.am
@@ -2,8 +2,15 @@
 
 install-exec-hook:
 	chmod +x boinc-client
-	$(INSTALL) -d $(DESTDIR)$(sysconfdir)/init.d
-	$(INSTALL) -b boinc-client $(DESTDIR)$(sysconfdir)/init.d/boinc-client
+	if [ -d /etc/init.d ] ; then \
+		$(INSTALL) -d $(DESTDIR)$(sysconfdir)/init.d ; \
+		$(INSTALL) -b boinc-client $(DESTDIR)$(sysconfdir)/init.d/boinc-client ; \
+	fi
+	if [ -d /usr/lib/systemd/system ] ; then \
+		$(INSTALL) -b boinc-client.service $(DESTDIR)/usr/lib/systemd/system/boinc-client.service ; \
+	elif [ -d /lib/systemd/system ] ; then \
+		$(INSTALL) -b boinc-client.service $(DESTDIR)/lib/systemd/system/boinc-client.service ; \
+	fi
 	if [ -d /etc/sysconfig ] ; then \
 	  $(INSTALL) -d $(DESTDIR)$(sysconfdir)/sysconfig ; \
 	  $(INSTALL) $(srcdir)/boinc-client.conf $(DESTDIR)$(sysconfdir)/sysconfig/boinc-client ; \
@@ -15,3 +22,6 @@ install-exec-hook:
 	  $(INSTALL) $(srcdir)/boinc-client.conf $(DESTDIR)$(sysconfdir)/boinc-client.conf ; \
 	fi
 
+clean:
+	rm boinc-client.service
+	rm boinc-client
diff --git a/client/scripts/boinc-client.service.in b/client/scripts/boinc-client.service.in
new file mode 100644
index 0000000000..2c424ffb5d
--- /dev/null
+++ b/client/scripts/boinc-client.service.in
@@ -0,0 +1,21 @@
+[Unit]
+Description=Berkeley Open Infrastructure Network Computing Client
+Documentation=man:boinc(1)
+After=network-online.target
+
+[Service]
+Type=simple
+Nice=19
+User=boinc
+WorkingDirectory=/var/lib/boinc-client
+EnvironmentFile=-/etc/sysconfig/boinc-client
+EnvironmentFile=-/etc/default/boinc-client
+EnvironmentFile=-/etc/boinc-client.conf
+ExecStart=@exec_prefix@/bin/boinc
+ExecStop=@exec_prefix@/bin/boinccmd --quit
+ExecReload=@exec_prefix@/bin/boinccmd --read_cc_config
+ExecStopPost=/bin/rm -f /var/lib/boinc-client/lockfile
+IOSchedulingClass=idle
+
+[Install]
+WantedBy=multi-user.target
diff --git a/configure.ac b/configure.ac
index 8e2115385c..d1a0654e61 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1318,6 +1318,7 @@ AC_CONFIG_FILES([
                  client/win/boinc_path_config.py:py/boinc_path_config.py.in
                  client/scripts/Makefile
                  client/scripts/boinc-client
+                 client/scripts/boinc-client.service
                  db/Makefile
                  doc/Makefile
                  doc/manpages/Makefile
